File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/exoplayer/preparation/mediasource/MediaSourceProvider.java
package com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.mediasource;

import android.net.Uri;

import com.google.android.exoplayer2.extractor.ExtractorsFactory;
import com.google.android.exoplayer2.extractor.mp3.Mp3Extractor;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.upstream.DefaultLoadErrorHandlingPolicy;
import com.google.android.exoplayer2.upstream.FileDataSourceFactory;
import com.google.android.exoplayer2.upstream.HttpDataSource;
import com.google.android.exoplayer2.upstream.cache.Cache;
import com.google.android.exoplayer2.upstream.cache.CacheDataSourceFactory;
import com.lasthopesoftware.bluewater.client.library.repository.Library;
import com.lasthopesoftware.bluewater.shared.IoCommon;
import com.namehillsoftware.lazyj.AbstractSynchronousLazy;
import com.namehillsoftware.lazyj.CreateAndHold;
import com.namehillsoftware.lazyj.Lazy;

public class MediaSourceProvider implements SpawnMediaSources {

	private static final CreateAndHold<ExtractorsFactory> extractorsFactory = new Lazy<>(() -> Mp3Extractor.FACTORY);

	private final CreateAndHold<ExtractorMediaSource.Factory> lazyFileExtractorFactory = new AbstractSynchronousLazy<ExtractorMediaSource.Factory>() {
		@Override
		protected ExtractorMediaSource.Factory create() {
			final ExtractorMediaSource.Factory factory = new ExtractorMediaSource.Factory(new FileDataSourceFactory());
			factory.setExtractorsFactory(extractorsFactory.getObject());
			return factory;
		}
	};

	private final CreateAndHold<ExtractorMediaSource.Factory> lazyRemoteExtractorFactory;

	public MediaSourceProvider(Library library, ProvideHttpDataSourceFactory dataSourceFactoryProvider, Cache cache) {

		lazyRemoteExtractorFactory = new AbstractSynchronousLazy<ExtractorMediaSource.Factory>() {
			@Override
			protected ExtractorMediaSource.Factory create() {
				final HttpDataSource.Factory httpDataSourceFactory = dataSourceFactoryProvider.getHttpDataSourceFactory(library);

				final CacheDataSourceFactory cacheDataSourceFactory = new CacheDataSourceFactory(
					cache,
					httpDataSourceFactory);

				final ExtractorMediaSource.Factory factory = new ExtractorMediaSource.Factory(cacheDataSourceFactory);
				factory.setLoadErrorHandlingPolicy(new DefaultLoadErrorHandlingPolicy(DefaultLoadErrorHandlingPolicy.DEFAULT_MIN_LOADABLE_RETRY_COUNT_PROGRESSIVE_LIVE));
				factory.setExtractorsFactory(extractorsFactory.getObject());
				return factory;
			}
		};
	}

	@Override
	public MediaSource getNewMediaSource(Uri uri) {
		return getFactory(uri).createMediaSource(uri);
	}

	private ExtractorMediaSource.Factory getFactory(Uri uri) {
		return IoCommon.FileUriScheme.equalsIgnoreCase(uri.getScheme())
			? lazyFileExtractorFactory.getObject()
			: lazyRemoteExtractorFactory.getObject();
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/exoplayer/preparation/mediasource/MediaSourceProvider.kt
package com.lasthopesoftware.bluewater.client.playback.file.exoplayer.preparation.mediasource

import android.content.Context
import android.net.Uri
import androidx.media3.common.MediaItem
import androidx.media3.common.util.UnstableApi
import androidx.media3.datasource.ContentDataSource
import androidx.media3.datasource.DataSource
import androidx.media3.datasource.FileDataSource
import androidx.media3.exoplayer.source.MediaSource
import androidx.media3.exoplayer.source.ProgressiveMediaSource
import androidx.media3.exoplayer.upstream.DefaultLoadErrorHandlingPolicy
import androidx.media3.extractor.DefaultExtractorsFactory
import androidx.media3.extractor.mp3.Mp3Extractor
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.promises.extensions.cancelBackThen
import com.lasthopesoftware.resources.uri.IoCommon
import com.namehillsoftware.handoff.promises.Promise

@UnstableApi class MediaSourceProvider(
	private val context: Context,
	private val remoteDataSourceFactory: ProvideRemoteDataSourceFactory
) : SpawnMediaSources {

	companion object {
		private val extractorsFactory by lazy { Mp3Extractor.FACTORY }

		private val promisedFileExtractorFactory by lazy {
			Promise<MediaSource.Factory>(ProgressiveMediaSource.Factory(FileDataSource.Factory(), extractorsFactory))
		}
	}

	private val promisedContentExtractorFactory by lazy {
		Promise<MediaSource.Factory>(ProgressiveMediaSource.Factory(ContentDataSourceFactory(context), DefaultExtractorsFactory()))
	}

	override fun promiseNewMediaSource(libraryId: LibraryId, uri: Uri): Promise<MediaSource> =
		getFactory(libraryId, uri).cancelBackThen { it, _ -> it.createMediaSource(MediaItem.Builder().setUri(uri).build()) }

	private fun getFactory(libraryId: LibraryId, uri: Uri): Promise<MediaSource.Factory> =
		when {
			IoCommon.fileUriScheme.equals(uri.scheme, ignoreCase = true) -> promisedFileExtractorFactory
			IoCommon.contentUriScheme.equals(uri.scheme, ignoreCase = true) -> promisedContentExtractorFactory
			else -> remoteDataSourceFactory
				.promiseRemoteDataSourceFactory(libraryId)
				.cancelBackThen { cacheDataSourceFactory, _ ->
					val factory = ProgressiveMediaSource.Factory(cacheDataSourceFactory, extractorsFactory)
					factory.setLoadErrorHandlingPolicy(
						DefaultLoadErrorHandlingPolicy(
							DefaultLoadErrorHandlingPolicy.DEFAULT_MIN_LOADABLE_RETRY_COUNT_PROGRESSIVE_LIVE
						)
					)
				}
		}

	private class ContentDataSourceFactory(private val context: Context) : DataSource.Factory {
		override fun createDataSource(): DataSource = ContentDataSource(context)
	}
}
